internal/graph/singleflight_resolver.go

```diff
@@ -0,0 +1,94 @@
+package graph
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/logger"
+)
+
+var (
+	deduplicatedDispatchesCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "deduplicated_dispatches_counter",
+		Help:      "The total number of calls to ResolveCheck that were deduplicated because of singleflight resolver.",
+	})
+
+	deduplicatedDBQueriesCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "deduplicated_db_queries_counter",
+		Help:      "The total number of DB queries that were deduplicated because of singleflight resolver.",
+	})
+)
+
+type singleflightCheckResolver struct {
+	delegate CheckResolver
+	group    singleflight.Group
+	logger   logger.Logger
+}
+
+func NewSingleflightCheckResolver(delegate CheckResolver) *singleflightCheckResolver {
```

Comment:
```suggestion
func NewSingleFlightCheckResolver(delegate CheckResolver) CheckResolver {
```

This exported function should not return an un-exported type, but the interface instead, as that will contain what the user can do with the struct returned from this exported func.